include/boost/math/ccmath/fma.hpp

```diff
@@ -0,0 +1,128 @@
+//  (C) Copyright Matt Borland 2022.
+//  Use, modification and distribution are subject to the
+//  Boost Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_CCMATH_FMA_HPP
+#define BOOST_MATH_CCMATH_FMA_HPP
+
+#include <cmath>
+#include <limits>
+#include <type_traits>
+#include <boost/math/tools/is_constant_evaluated.hpp>
+#include <boost/math/ccmath/isinf.hpp>
+#include <boost/math/ccmath/isnan.hpp>
+
+namespace boost::math::ccmath {
+
+namespace detail {
+
+template <typename T>
+constexpr T fma_imp(const T x, const T y, const T z) noexcept
+{
+    #if defined(__GNUC__) && !defined(__clang__) && !defined(__INTEL_COMPILER) && !defined(__INTEL_LLVM_COMPILER)
+    if constexpr (std::is_same_v<T, float>)
+    {
+        return __builtin_fmaf(x, y, z);
+    }
+    else if constexpr (std::is_same_v<T, double>)
+    {
+        return __builtin_fma(x, y, z);
+    }
+    else if constexpr (std::is_same_v<T, long double>)
+    {
+        return __builtin_fmal(x, y, z);
+    }
+    #endif
+    
+    // If we can't use compiler intrinsics hope that -fma flag optimizes this call to fma instruction
+    return (x * y) + z;
```

Comment:
So std::fma won't work here?

Also, why not just place the asm directly in the code if we detect x86 >= ivy bridge?

Reply:
`std::fma` is likely not constexpr until C++23. [According to this accepted C++20 proposal](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1668r1.html) inlining asm is not needed if `std::is_constant_evaluated` is true